// ng-keyboard-shortcuts/esm2015/lib/ng-keyboard-shortcuts.module.js
import * as tslib_1 from "tslib";
var KeyboardShortcutsModule_1;
import { NgModule } from "@angular/core";
import { KeyboardShortcutsComponent } from "./ng-keyboard-shortcuts.component";
import { KeyboardShortcutsService } from "./ng-keyboard-shortcuts.service";
import { KeyboardShortcutsHelpService } from './ng-keyboard-shortcuts-help.service';
import { KeyboardShortcutsDirective } from './ng-keyboard-shortcuts.directive';
import { KeyboardShortcutsHelpComponent } from './ng-keyboard-shortcuts-help.component';
import { KeyboardShortcutsHelpItemComponent } from './ng-keyboard-shortcuts-help-item.component';
import { CommonModule } from "@angular/common";
let KeyboardShortcutsModule = KeyboardShortcutsModule_1 = class KeyboardShortcutsModule {
    static forRoot() {
        return {
            ngModule: KeyboardShortcutsModule_1,
            providers: [
                KeyboardShortcutsService,
                KeyboardShortcutsHelpService
            ]
        };
    }
};
KeyboardShortcutsModule = KeyboardShortcutsModule_1 = tslib_1.__decorate([
    NgModule({
        imports: [CommonModule],
        entryComponents: [KeyboardShortcutsHelpComponent],
        declarations: [KeyboardShortcutsComponent, KeyboardShortcutsDirective, KeyboardShortcutsHelpComponent, KeyboardShortcutsHelpItemComponent],
        exports: [KeyboardShortcutsComponent, KeyboardShortcutsDirective, KeyboardShortcutsHelpComponent]
    })
], KeyboardShortcutsModule);
export { KeyboardShortcutsModule };
