// ng-keyboard-shortcuts/esm2015/lib/ng-keyboard-shortcuts.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from "@angular/core";
import { KeyboardShortcutsService } from "./ng-keyboard-shortcuts.service";
/**
 * A component to bind global shortcuts, can be used multiple times across the app
 * will remove registered shortcuts when element is removed from DOM.
 */
let KeyboardShortcutsComponent = class KeyboardShortcutsComponent {
    /**
     * @ignore
     * @param {KeyboardShortcutsService} keyboard
     */
    constructor(keyboard) {
        this.keyboard = keyboard;
        /**
         * A list of shortcuts.
         */
        this.shortcuts = [];
        /**
         * @ignore
         * list of registered keyboard shortcuts
         * used for clean up on NgDestroy.
         */
        this.clearIds = [];
        /**
         * @ignore
         */
        this._disabled = false;
    }
    /**
     * Disable all shortcuts for this component.
     */
    set disabled(value) {
        this._disabled = value;
        if (this.clearIds) {
            this.keyboard.remove(this.clearIds);
            this.clearIds = [];
        }
        if (value) {
            return;
        }
        this.clearIds = this.keyboard.add(this.shortcuts);
    }
    /**
     * @ignore
     */
    ngOnInit() { }
    /**
     * Select a key to listen to, will emit when the selected key is pressed.
     */
    select(key) {
        return this.keyboard.select(key);
    }
    /**
     * @ignore
     */
    ngAfterViewInit() { }
    /**
     * @ignore
     */
    ngOnChanges(changes) {
        if (!changes.shortcuts || !changes.shortcuts.currentValue) {
            return;
        }
        if (this.clearIds) {
            this.keyboard.remove(this.clearIds);
        }
        setTimeout(() => this.clearIds = this.keyboard.add(changes.shortcuts.currentValue));
    }
    /**
     * @ignore
     */
    ngOnDestroy() {
        this.keyboard.remove(this.clearIds);
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], KeyboardShortcutsComponent.prototype, "shortcuts", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], KeyboardShortcutsComponent.prototype, "disabled", null);
KeyboardShortcutsComponent = tslib_1.__decorate([
    Component({
        selector: "ng-keyboard-shortcuts",
        template: ""
    }),
    tslib_1.__metadata("design:paramtypes", [KeyboardShortcutsService])
], KeyboardShortcutsComponent);
export { KeyboardShortcutsComponent };
