// ng-keyboard-shortcuts/esm2015/lib/ng-keyboard-shortcuts.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input } from "@angular/core";
import { AllowIn } from "./ng-keyboard-shortcuts.interfaces";
import { KeyboardShortcutsService } from "./ng-keyboard-shortcuts.service";
/**
 * A directive to be used with "focusable" elements, like:
 * textarea, input, select.
 */
let KeyboardShortcutsDirective = class KeyboardShortcutsDirective {
    /**
     * @ignore
     * @param {KeyboardShortcutsService} keyboard
     * @param {ElementRef} el
     */
    constructor(keyboard, el) {
        this.keyboard = keyboard;
        this.el = el;
        /**
         * @ignore
         * @type {boolean}
         * @private
         */
        this._disabled = false;
    }
    /**
     * whether to disable the shortcuts for this directive
     * @param value
     */
    set disabled(value) {
        this._disabled = value;
        if (this.clearIds) {
            this.keyboard.remove(this.clearIds);
        }
        setTimeout(() => {
            if (value === false && this.ngKeyboardShortcuts) {
                this.clearIds = this.keyboard.add(this.transformInput(this.ngKeyboardShortcuts));
            }
        });
    }
    /**
     * @ignore
     * @param {Shortcut[]} shortcuts
     * @returns {any}
     */
    transformInput(shortcuts) {
        return shortcuts.map(shortcut => (Object.assign({}, shortcut, { target: this.el.nativeElement, allowIn: [AllowIn.Select, AllowIn.Input, AllowIn.Textarea] })));
    }
    /**
     * @ignore
     */
    ngOnDestroy() {
        if (!this.clearIds) {
            return;
        }
        this.keyboard.remove(this.clearIds);
    }
    /**
     * @ignore
     * @param {SimpleChanges} changes
     */
    ngOnChanges(changes) {
        const { ngKeyboardShortcuts } = changes;
        if (this.clearIds) {
            this.keyboard.remove(this.clearIds);
        }
        if (!ngKeyboardShortcuts || !ngKeyboardShortcuts.currentValue) {
            return;
        }
        this.clearIds = this.keyboard.add(this.transformInput(ngKeyboardShortcuts.currentValue));
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array)
], KeyboardShortcutsDirective.prototype, "ngKeyboardShortcuts", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], KeyboardShortcutsDirective.prototype, "disabled", null);
KeyboardShortcutsDirective = tslib_1.__decorate([
    Directive({
        selector: "[ngKeyboardShortcuts]"
    }),
    tslib_1.__metadata("design:paramtypes", [KeyboardShortcutsService, ElementRef])
], KeyboardShortcutsDirective);
export { KeyboardShortcutsDirective };
