// ng-keyboard-shortcuts/esm2015/lib/ng-keyboard-shortcuts-help-item.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from "@angular/core";
import { symbols } from "./keys";
import { identity } from './utils';
/**
 * @ignore
 */
let KeyboardShortcutsHelpItemComponent = class KeyboardShortcutsHelpItemComponent {
    constructor() { }
    set shortcut(shortcut) {
        const key = Array.isArray(shortcut.key) ? shortcut.key : [shortcut.key];
        this.parsedKeys = key.map(key => key
            .split(" ")
            .filter(identity)
            .filter(key => key !== "+")
            .map(key => {
            if (symbols[key]) {
                return symbols[key];
            }
            return key;
        }));
        this._shortcut = shortcut;
    }
    get shortcut() {
        return this._shortcut;
    }
    ngOnInit() { }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Number)
], KeyboardShortcutsHelpItemComponent.prototype, "index", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], KeyboardShortcutsHelpItemComponent.prototype, "shortcut", null);
KeyboardShortcutsHelpItemComponent = tslib_1.__decorate([
    Component({
        selector: "ng-keyboard-shortcuts-help-item",
        template: "<div class=\"item\" [class.item--odd]=\"index % 2 !== 0\" *ngIf=\"shortcut.description\">\n  <div class=\"description\">\n    <span>{{shortcut.description}}</span>\n  </div>\n  <div class=\"keys\">\n    <div *ngFor=\"let sKey of parsedKeys;let i = index\" class=\"key__container\">\n      <span class=\"key\" *ngFor=\"let key of sKey;\">{{key}}</span>\n      <span *ngIf=\"parsedKeys.length > 1 && i < parsedKeys.length - 1\" class=\"separator\"> / </span>\n    </div>\n  </div>\n</div>\n",
        styles: [".key{border:1px solid #ccc;border-radius:4px;padding:5px 12px;margin-right:5px;background-color:#f5f5f5}.key__container{display:inline-block}.separator{margin-right:5px}.keys{float:right}.item{background-color:#ebebeb;padding:12px}.description{min-width:168px;display:inline-block;color:#333}.item--odd{background-color:#fff}"]
    }),
    tslib_1.__metadata("design:paramtypes", [])
], KeyboardShortcutsHelpItemComponent);
export { KeyboardShortcutsHelpItemComponent };
