// ng-keyboard-shortcuts/esm5/lib/ng-keyboard-shortcuts-help-item.component.js
import * as tslib_1 from "tslib";
import { Component, Input } from "@angular/core";
import { symbols } from "./keys";
import { identity } from './utils';
/**
 * @ignore
 */
var KeyboardShortcutsHelpItemComponent = /** @class */ (function () {
    function KeyboardShortcutsHelpItemComponent() {
    }
    Object.defineProperty(KeyboardShortcutsHelpItemComponent.prototype, "shortcut", {
        get: function () {
            return this._shortcut;
        },
        set: function (shortcut) {
            var key = Array.isArray(shortcut.key) ? shortcut.key : [shortcut.key];
            this.parsedKeys = key.map(function (key) {
                return key
                    .split(" ")
                    .filter(identity)
                    .filter(function (key) { return key !== "+"; })
                    .map(function (key) {
                    if (symbols[key]) {
                        return symbols[key];
                    }
                    return key;
                });
            });
            this._shortcut = shortcut;
        },
        enumerable: true,
        configurable: true
    });
    KeyboardShortcutsHelpItemComponent.prototype.ngOnInit = function () { };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Number)
    ], KeyboardShortcutsHelpItemComponent.prototype, "index", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], KeyboardShortcutsHelpItemComponent.prototype, "shortcut", null);
    KeyboardShortcutsHelpItemComponent = tslib_1.__decorate([
        Component({
            selector: "ng-keyboard-shortcuts-help-item",
            template: "<div class=\"item\" [class.item--odd]=\"index % 2 !== 0\" *ngIf=\"shortcut.description\">\n  <div class=\"description\">\n    <span>{{shortcut.description}}</span>\n  </div>\n  <div class=\"keys\">\n    <div *ngFor=\"let sKey of parsedKeys;let i = index\" class=\"key__container\">\n      <span class=\"key\" *ngFor=\"let key of sKey;\">{{key}}</span>\n      <span *ngIf=\"parsedKeys.length > 1 && i < parsedKeys.length - 1\" class=\"separator\"> / </span>\n    </div>\n  </div>\n</div>\n",
            styles: [".key{border:1px solid #ccc;border-radius:4px;padding:5px 12px;margin-right:5px;background-color:#f5f5f5}.key__container{display:inline-block}.separator{margin-right:5px}.keys{float:right}.item{background-color:#ebebeb;padding:12px}.description{min-width:168px;display:inline-block;color:#333}.item--odd{background-color:#fff}"]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], KeyboardShortcutsHelpItemComponent);
    return KeyboardShortcutsHelpItemComponent;
}());
export { KeyboardShortcutsHelpItemComponent };
