// ng-keyboard-shortcuts/esm5/lib/ng-keyboard-shortcuts.module.js
import * as tslib_1 from "tslib";
import { NgModule } from "@angular/core";
import { KeyboardShortcutsComponent } from "./ng-keyboard-shortcuts.component";
import { KeyboardShortcutsService } from "./ng-keyboard-shortcuts.service";
import { KeyboardShortcutsHelpService } from './ng-keyboard-shortcuts-help.service';
import { KeyboardShortcutsDirective } from './ng-keyboard-shortcuts.directive';
import { KeyboardShortcutsHelpComponent } from './ng-keyboard-shortcuts-help.component';
import { KeyboardShortcutsHelpItemComponent } from './ng-keyboard-shortcuts-help-item.component';
import { CommonModule } from "@angular/common";
var KeyboardShortcutsModule = /** @class */ (function () {
    function KeyboardShortcutsModule() {
    }
    KeyboardShortcutsModule_1 = KeyboardShortcutsModule;
    KeyboardShortcutsModule.forRoot = function () {
        return {
            ngModule: KeyboardShortcutsModule_1,
            providers: [
                KeyboardShortcutsService,
                KeyboardShortcutsHelpService
            ]
        };
    };
    var KeyboardShortcutsModule_1;
    KeyboardShortcutsModule = KeyboardShortcutsModule_1 = tslib_1.__decorate([
        NgModule({
            imports: [CommonModule],
            entryComponents: [KeyboardShortcutsHelpComponent],
            declarations: [KeyboardShortcutsComponent, KeyboardShortcutsDirective, KeyboardShortcutsHelpComponent, KeyboardShortcutsHelpItemComponent],
            exports: [KeyboardShortcutsComponent, KeyboardShortcutsDirective, KeyboardShortcutsHelpComponent]
        })
    ], KeyboardShortcutsModule);
    return KeyboardShortcutsModule;
}());
export { KeyboardShortcutsModule };
