// ng-keyboard-shortcuts/esm5/lib/ng-keyboard-shortcuts.directive.js
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input } from "@angular/core";
import { AllowIn } from "./ng-keyboard-shortcuts.interfaces";
import { KeyboardShortcutsService } from "./ng-keyboard-shortcuts.service";
/**
 * A directive to be used with "focusable" elements, like:
 * textarea, input, select.
 */
var KeyboardShortcutsDirective = /** @class */ (function () {
    /**
     * @ignore
     * @param {KeyboardShortcutsService} keyboard
     * @param {ElementRef} el
     */
    function KeyboardShortcutsDirective(keyboard, el) {
        this.keyboard = keyboard;
        this.el = el;
        /**
         * @ignore
         * @type {boolean}
         * @private
         */
        this._disabled = false;
    }
    Object.defineProperty(KeyboardShortcutsDirective.prototype, "disabled", {
        /**
         * whether to disable the shortcuts for this directive
         * @param value
         */
        set: function (value) {
            var _this = this;
            this._disabled = value;
            if (this.clearIds) {
                this.keyboard.remove(this.clearIds);
            }
            setTimeout(function () {
                if (value === false && _this.ngKeyboardShortcuts) {
                    _this.clearIds = _this.keyboard.add(_this.transformInput(_this.ngKeyboardShortcuts));
                }
            });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @ignore
     * @param {Shortcut[]} shortcuts
     * @returns {any}
     */
    KeyboardShortcutsDirective.prototype.transformInput = function (shortcuts) {
        var _this = this;
        return shortcuts.map(function (shortcut) { return (tslib_1.__assign({}, shortcut, { target: _this.el.nativeElement, allowIn: [AllowIn.Select, AllowIn.Input, AllowIn.Textarea] })); });
    };
    /**
     * @ignore
     */
    KeyboardShortcutsDirective.prototype.ngOnDestroy = function () {
        if (!this.clearIds) {
            return;
        }
        this.keyboard.remove(this.clearIds);
    };
    /**
     * @ignore
     * @param {SimpleChanges} changes
     */
    KeyboardShortcutsDirective.prototype.ngOnChanges = function (changes) {
        var ngKeyboardShortcuts = changes.ngKeyboardShortcuts;
        if (this.clearIds) {
            this.keyboard.remove(this.clearIds);
        }
        if (!ngKeyboardShortcuts || !ngKeyboardShortcuts.currentValue) {
            return;
        }
        this.clearIds = this.keyboard.add(this.transformInput(ngKeyboardShortcuts.currentValue));
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Array)
    ], KeyboardShortcutsDirective.prototype, "ngKeyboardShortcuts", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], KeyboardShortcutsDirective.prototype, "disabled", null);
    KeyboardShortcutsDirective = tslib_1.__decorate([
        Directive({
            selector: "[ngKeyboardShortcuts]"
        }),
        tslib_1.__metadata("design:paramtypes", [KeyboardShortcutsService, ElementRef])
    ], KeyboardShortcutsDirective);
    return KeyboardShortcutsDirective;
}());
export { KeyboardShortcutsDirective };
