// ng-keyboard-shortcuts/esm5/lib/ng-keyboard-shortcuts-help.service.js
import * as tslib_1 from "tslib";
import { Injectable } from "@angular/core";
import { KeyboardShortcutsService } from "./ng-keyboard-shortcuts.service";
import { map } from "rxjs/operators";
import * as i0 from "@angular/core";
import * as i1 from "./ng-keyboard-shortcuts.service";
/**
 * Service to assist showing custom help screen
 */
var KeyboardShortcutsHelpService = /** @class */ (function () {
    /**
     * @ignore
     * @param {KeyboardShortcutsService} keyboard
     */
    function KeyboardShortcutsHelpService(keyboard) {
        this.keyboard = keyboard;
        /**
         * Observable to provide access to all registered shortcuts in the app.
         * @type {Observable<any>}
         */
        this.shortcuts$ = this.keyboard.shortcuts$.pipe(map(function (shortcuts) {
            return shortcuts
                .filter(function (shortcut) { return Boolean(shortcut.label) && Boolean(shortcut.description); })
                .map(function (_a) {
                var key = _a.key, label = _a.label, description = _a.description;
                return ({
                    key: key,
                    label: label,
                    description: description
                });
            });
        }));
    }
    KeyboardShortcutsHelpService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function KeyboardShortcutsHelpService_Factory() { return new KeyboardShortcutsHelpService(i0.ɵɵinject(i1.KeyboardShortcutsService)); }, token: KeyboardShortcutsHelpService, providedIn: "root" });
    KeyboardShortcutsHelpService = tslib_1.__decorate([
        Injectable({
            providedIn: "root"
        }),
        tslib_1.__metadata("design:paramtypes", [KeyboardShortcutsService])
    ], KeyboardShortcutsHelpService);
    return KeyboardShortcutsHelpService;
}());
export { KeyboardShortcutsHelpService };
