// ng-keyboard-shortcuts/esm5/lib/utils.js
var _this = this;
import * as tslib_1 from "tslib";
export function isFunction(x) {
    return typeof x === "function";
}
export var any = function (fn, list) {
    var idx = 0;
    while (idx < list.length) {
        if (fn(list[idx])) {
            return true;
        }
        idx += 1;
    }
    return false;
};
export var identity = function (x) { return x; };
/**
 * @ignore
 * @param x
 * @returns boolean
 */
export var isNill = function (x) { return x == null; };
/**
 * @ignore
 * @param xs
 * @param key
 * @returns any
 */
export var groupBy = function (xs, key) {
    return xs.reduce(function (result, x) {
        var _a;
        return (tslib_1.__assign({}, result, (_a = {}, _a[x[key]] = tslib_1.__spread((result[x[key]] || []), [x]), _a)));
    }, {});
};
/**
 * @ignore
 * @param first
 * @param second
 * @returns any[]
 */
export var difference = function (first, second) {
    return first.filter(function (item) { return !second.includes(item); });
};
/**
 * @ignore
 * @param preds
 * @returns (...args) => boolean;
 */
export var allPass = function (preds) { return function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    var idx = 0;
    var len = preds.length;
    while (idx < len) {
        if (!preds[idx].apply(_this, args)) {
            return false;
        }
        idx += 1;
    }
    return true;
}; };
export var prop = function (prop) { return function (object) { return object[prop]; }; };
var minMaxArrayProp = function (type) { return function (property, array) {
    return Math[type].apply(Math, array.map(prop(property)));
}; };
var ɵ0 = minMaxArrayProp;
export var maxArrayProp = function (property, array) {
    var _a;
    return array.reduce(function (acc, curr) {
        var propFn = prop(property);
        var currentValue = propFn(curr);
        var previousValue = propFn(acc);
        return currentValue > previousValue ? curr : acc;
    }, (_a = {}, _a[property] = 0, _a));
};
export { ɵ0 };
